petastorm/pytorch.py

```diff
@@ -423,3 +367,139 @@ def __enter__(self):
+    return (keys, buffer)
+
+
+class InMemBatchedDataLoader(object):
+    """
+    Same as BatchedDataLoader except it only loads upto capacity rows into memory.
+    This class doesn't allow to be used multiple-times, so it doesn't inherit LoaderBase.
+    """
+
+    def __init__(self, reader, batch_size=1,
+                 transform_fn=None,
+                 num_epochs=1,
+                 seed=0,
+                 rows_capacity=1024,
+                 shuffle=False):
+        """
+        Initializes a data loader object.
+        This class does not support special types that are not supported in PyTorch (decimal/string).
+        :param reader: petastorm Reader instance
+        :param batch_size: the number of items to return per batch; factored into the len() of this reader
+        :param transform_fn: an optional callable to convert batches from the reader to PyTorch tensors
+        :param num_epochs: number of epochs.
+        :param seed: random seed used to shuffle.
+        :param rows_capacity: number of rows to be loaded into memory.
+        :param shuffle: If ``True``, indices will be shuffled in every epoch.
+        """
+        super(InMemBatchedDataLoader, self).__init__()
+        self._reader = reader
+        self._batch_size = batch_size
+        self._transform_fn = transform_fn or torch.as_tensor
+        self._num_epochs = num_epochs
+        self._seed = seed
+        self._rows_capacity = rows_capacity
+        self._shuffle = shuffle
+        self._in_iter = False
+
+    def __iter__(self):
+        """
+        The Data Loader iterator stops the for-loop when num_epochs is reached.
+        """
+        if self._in_iter:
+            raise RuntimeError("InMemBatchedDataLoader couldn't be used multiple times, please\
+                    specify total number of epochs using num_epochs in constructor.")
+        self._in_iter = True
+        try:
+            keys, buffer = _load_rows_into_mem(self._reader, self._transform_fn, self._rows_capacity)
+            for epoch in range(self._num_epochs):
+                size = len(buffer[0])
+                if self._shuffle:
+                    # Deterministically shuffle based on seed and current epoch id.
+                    g = torch.Generator()
+                    g.manual_seed(self._seed + epoch)
+                    indices = torch.randperm(size, generator=g).tolist()
+                else:
+                    indices = list(range(size))
+
+                # Sample batches
+                for i in range(0, len(indices), self._batch_size):
+                    idx = indices[i:i+self._batch_size]
+                    batch = [v[idx] for v in buffer]
+                    size -= len(batch[0])
+                    batch = dict(zip(keys, batch))
+                    yield batch
+                assert size == 0
+        except Exception as e:
+            logger.error('Iteration on InMemBatchedDataLoader raise error: %s', repr(e))
```

Comment:
What is the added value of this generic exception handler?

Reply:
I'm not sure if `try` block should be used TBH. I was seeing the `LoaderBase` having `try` block for `__iter__`. I feel the only exception may come from `_load_rows_into_mem ` with OOM or reader corruption? Any suggestions?

Reply:
It could also come from within reader errors. In LoaderBase, this PR: #544 added the try/except and the goal was to raise an explicit error in the second iteration. I think the goal was to support multiple `iter(data_loader)` calls, but fail the next time if previous iterations ended up with an exception.

Reply:
Hmm, after a second thought, I decide remove `try` block since `_in_iter` is already checked and raised if possible.